File: Sheets/src/neetcode/practice/Stack/_5_dailyTemperatures.java
package neetcode.practice.Stack;

import java.util.AbstractMap;
import java.util.ArrayDeque;
import java.util.Deque;

public class _5_dailyTemperatures {
  public record Pair<K, V>(K key, V value) {}

  public static int[] dailyTemperatures(int[] temperatures) {
    int n = temperatures.length;
    int[] ans = new int[n];

    Deque<Pair<Integer, Integer>> stack = new ArrayDeque<>();

    for (int i = 0; i < n; i++) {
      int t = temperatures[i];

      while (!stack.isEmpty() && t > stack.peek().key()) {
        Pair<Integer, Integer> pair = stack.pop();
        ans[pair.value()] = i - pair.value();
      }

      stack.push(new Pair<>(t, i));
    }

    return ans;
  }

  public static int[] dailyTemperaturesAlternate(int[] temperatures) {
    int n = temperatures.length;
    int[] ans = new int[n];

    Deque<AbstractMap.SimpleEntry<Integer, Integer>> stack = new ArrayDeque<>();

    for (int i = 0; i < n; i++) {
      int t = temperatures[i];

      while (!stack.isEmpty() && t > stack.peek().getKey()) {
        AbstractMap.SimpleEntry<Integer, Integer> pair = stack.pop();
        ans[pair.getValue()] = i - pair.getValue();
      }

      stack.push(new AbstractMap.SimpleEntry<>(t, i));
    }

    return ans;
  }
}

File: Sheets/src/neetcode/practice/Stack/_5_dailyTemperatures.cc
#include <stack>
#include <vector>

class Solution {
public:
  std::vector<int> dailyTemperatures(std::vector<int> &temperatures) {
    int n = static_cast<int>(temperatures.size());

    std::stack<std::pair<int, int>> stk{};
    std::vector<int> res(n);
    for (int i = n - 1; i >= 0; i--) {
      int val = temperatures[i];

      if (stk.empty() || stk.top().first <= val) {
        while (!stk.empty() && stk.top().first <= val)
          stk.pop();
      }

      res[i] = (stk.empty()) ? 0 : stk.top().second - i;
      stk.push({val, i});
    }

    return res;
  }
};
